### DISCUSSION TITLE
What happens if FontName has comma in it?

### DISCUSSION THREAD

**Discussion by p0358** [libass/libass#842] [Q&A] [answered]
Just curious is all, currently the parser doesn't seem to take it into account at all:
https://github.com/libass/libass/blob/6a759836e5e76bb7b69b0ac244eea76b0d290512/libass/ass.c#L425-L452

(I noticed that styles override parser supports escaping commas, but that's a separate thing)

Well the easy answer I guess is then it can't have a comma or parsing will work, right? But is it a wontfix and problem of someone creating such a font or would standard for ASS file specify some escaping that isn't done in libass?

Wondering what would happen if programs like Aegisub were to save a file specifying such a font, would it try to escape it or just produce a broken file with too many commas in line?

It's really not important by any means, just in my little project of using libass to get font info I also made a query path of `Fonts/{family},{bold},{italic}` and then thought whether I may need to escape `{family}`. Seems like I don't need to, but also might just end up replacing `,{bold},{italic}` with query params instead ¯\_(ツ)_/¯

**Comment by astiob** [comment, marked as answer]
To clarify for other readers: this is about font names in `Style` declarations. `\fnfoo,bar` works as usual in override tags. (But `\fn(foo,bar)`, of course, fails.)

> would standard for ASS file specify some escaping that isn't done in libass?

No.

> is it a wontfix and problem of someone creating such a font

Yes. Or rather, of someone who wants to use such a font in ASS: they’ll have to use inline override tags or edit the font file. In a similar but reverse manner, ASS doesn’t support font names with trailing spaces in inline override tags but does support them in style declarations. (And by the way: we have, in fact, seen one real font with a trailing space in its name, just this May.) Meanwhile, font names with _leading_ spaces aren’t supported in either location and are completely unusable.

> Wondering what would happen if programs like Aegisub were to save a file specifying such a font

I’ve just checked out of curiosity, and it seems Aegisub silently replaces the comma by a semicolon when the style properties dialog is closed.

**Reply by p0358** [reply]
Thanks. So I guess I won't worry about escaping such fonts too much then.

> we have, in fact, seen one real font with a trailing space in its name, just this May

I did a very short glance at my collection and saw a bunch of double spaces inside (between words), so it seems various weird typographic mistakes can happen lol

> I’ve just checked out of curiosity, and it seems Aegisub silently replaces the comma by a semicolon when the style properties dialog is closed.

Hm, I guess the only way to ever support them in something like libass then later on would be to replace all commas with semicolons in font names that fontcache picks up from the system into its cache. Can't "unescape" what's loaded from ASS since there could be a legit semicolon in font name too.

In practice probably not worth doing anything about it. Though I could totally imagine someone naming a font like `Some Cool Font, Improved Version` and not realize what a time bomb they're setting lol

**Reply by astiob** [reply]
I may be wrong, but it sounds to me like you think the font works in Aegisub but not in libass? It doesn’t. Aegisub uses libass or VSFilter for rendering, so it can’t make anything extra work that we don’t. When it does replace the comma by a semicolon, the user immediately sees in the output rendering that the desired font isn’t actually being used.

**Reply by p0358** [reply]
Ah so that's how it is. Okay, then zero problem since nobody will release broken subtitles without knowing in such case.

**Comment by astiob** [comment]
> (I noticed that styles override parser supports escaping commas, but that's a separate thing)

By the way, what do you mean by this? Our style overrides [are tokenized](https://github.com/libass/libass/blob/6a759836e5e76bb7b69b0ac244eea76b0d290512/libass/ass.c#L523-L612) the same way as normal styles, using the same `advance_token_pos`, and as far as I’m aware, there’s no escape mechanism.

**Reply by p0358** [reply]
I was basing that on this comment: https://github.com/libass/libass/blob/6a759836e5e76bb7b69b0ac244eea76b0d290512/libass/ass_parse.c#L318-L321

at the top of func it says `\brief Parse style override tags.`

**Reply by astiob** [reply]
OK, I see what’s happening now. I’m afraid you’re considerably misunderstanding this, so let me explain.

ASS has “styles”:
```
Style: Default,Arial,96,&H000000FF,&H80FFFF00,&H00000000,&H0000FF00,0,0,0,0,100,100,0,0,1,0,0,2,0,10,20,1
```

and “override tags” (or just “tags”), which are put inside lines of text (labelled `Dialogue` in the format):
```
Dialogue: 0,0:00:00.00,0:00:10.00,JP3,,0,0,0,,{\bord5\3a&H80\fnfoo}ABC DEF {\kf300}GHI{\fnbar}DEF {\kf300}broken\Nline
```

The entirety of `ass_parse.c` deals exclusively with dialogue text. The function you reference above parses override tags. The comment refers to tags like this:

```
Dialogue: 0,0:00:00.00,0:00:10.00,JP2,,0,0,0,,{\fs100\fscx0\t(0,1000,\pos(400,400)}scratch
```

Here, `\t` has three arguments: `0`, `1000`, and `\pos(400,400`, which contains a comma. Note that the backslash itself is not stripped. Even here, it’s not an escape character.

Finally, libass has a user style override API, which is not part of ASS proper. It is exposed directly to end users in mpv’s [`--sub-ass-style-overrides`](https://mpv.io/manual/stable/#options-sub-ass-style-overrides) option (and other players are welcome to do similarly). These are “overrides” in the sense that they take priority over the values specified in the ASS stream, and they’re “style” overrides in the sense that they target ASS style definitions and still have a lower priority than inline tags. When we talk about “style overrides”, in the context of libass, we usually mean these.

**Reply by p0358** [reply]
Okay that makes sense, just special nested handling of comma for arguments rather than escaping of comma

> `Dialogue: 0,0:00:00.00,0:00:10.00,JP2,,0,0,0,,{\fs100\fscx0\t(0,1000,\pos(400,400)}scratch`

Is there meant to be just a single closing `)` or is that a typo?

**Reply by astiob** [reply]
Meant in this case, but FWIW if you ever author your own ASS files for public distribution, please do use double closing `)` in cases like this :-)

This is a quirk of the format: you’ll notice I said the last argument is parsed as `\pos(400,400` without the closing parenthesis. There’s no special handling of nested parentheses, so the nested argument can never contain the closing paren. But the closing paren isn’t required in the first place, so `{\t(0,1000,\pos(400,400))}` is the same as `{\t(0,1000,\pos(400,400)}` and indeed the same as `{\t(0,1000,\pos(400,400}`.

_Except_ the internal subtitle renderers of MPC-HC and its derivatives (as well as VSFilterMod, from which they inherit the relevant code) _do_ have special handling of nested parentheses, so in these renderers only, `\t(0,1000,\pos(400,400)\fs20` and `\t(0,1000,\pos(400,400))\fs20` mean different things, so portable files should always include the second closing paren if any other tags follow.

**Reply by p0358** [reply]
Fair. And parsing quirks like this is exactly why it's probably desirable to prefer a widely-used parser over something whipped up on the side (unless it's a copy-paste of original parser logic, but then it *is* that parser logically as well still).

I was thinking that it actually wouldn't be that hard to make a minimal renderer by only making it aware of what tags are complex and special nested handling of just \t nested params. As I think technically the format of tags is otherwise well-defined and they could be dumbly skipped over without knowing what they do and their amount of args etc.

But then I took a closer look into ass_parse_tags and actually `RenderContext` and `ass_parse_tags` could easily be fixed to be just pure dry state processing without renderer. Most references to renderer are just erroneous (like doing `render_priv->library` instead of `state->library). Then the only changes would be to remove `renderer`, `shaper`, `rasterizer` (latter two seem unused), add `track` directly to context and move `time` to context as the comment in one place suggests: `// FIXME: move to render_context`.

And then for my parsing I could just reimplement minimal `parse_events` from ass_render.c that'd only go through tags and trigger font lookup if there's any text between them more-less (apart from \p drawing mode)
